_posts/2020-02-09-single-responsibility-principle.md

```diff
@@ -0,0 +1,133 @@
+ - The responsibility itself (i. e. what does the component actually do)
+ - The non-functional requirements to the application or the component we're developing
+ - How long we plan to support the code in the future
+ - How many people will work with this code
+ - Etc.
+
+However, this shouldn't intimidate us. **We should just split (or merge) components while we see that the certain software quality aspects keep improving.**
+
+Thus, the next step is to analyze how the Single Responsibility Principle affects the specific software qualities.
+
+### What software qualities does Single Responsibility Principle impact?
+
+#### Understandability and Learning Curve
+When we split responsibilities between smaller methods and classes, usually the system becomes easier to learn overall. We can learn bite-sized components one at a time, iteratively. When we jump into a new codebase, we can learn fine-grained components as we need them, ignoring the internals of the other components which are not yet relevant for us.
+
+If you have ever worked with code where the Single Responsibility Principle was not regarded much, you probably remember the frustration when you stumble upon a three hundred-line method or a thousand-line class about which you need to understand something, probably a little thing, but in order to figure that out you are forced to read through the whole method or the class. This not only takes a lot of time and mental energy, but also fills the "memory cache" of your brain with junk information that is completely irrelevant at the moment.
+
+It is possible, however, to take the separation of concerns so far that it might become harder to understand the logic. Returning to the `processTransaction()` example, consider the following way of implementing it:
+
+```
+class TransactionProcessor {
+    private final TransactionInstrumentation instrumentation;
+    
+    ...
+    
+    void processTransaction(Transaction t) {
+      if (isFraud(t)) {
+        instrumentation.detectedFraud(t);
+      }
+    }
+}
+
+class TransactionInstrumentation {
+  private final Logger logger;
+  private final AlertingService alertingService;
+
+  ...
+
+  void detectedFraud(Transaction t) {
+    logger.log("Detected fraud transaction {}", t);
+    alertingService.sendAlert(new FraudTransactionAlert(t));
+  }
+}
+```
+This approach is not unreasonable. Compared to the original version, it aids the *flexibility* and the *testability* of the code, as we will discuss below in this article. (In fact, I took the idea directly from the excellent article about [domain-oriented observability](https://martinfowler.com/articles/domain-oriented-observability.html) by Pete Hodgson.)
+
+On the other hand, it smears the logic so thin across multiple classes and methods that it would take longer to learn it than the original, at least for me. [1]
```

Comment:
The blog doesn't support foot notes. It does support "notice boxes" though (search for `<div class="notice success">` in the blog codebase). Perhaps that's an alternative.

Reply:
Notice boxes emphasize. I want to move this part out of the article flow. Maybe asterisk with `<abbr>`-like popup?